### PR TITLE
Add tsubasa keyboard

### PR DESCRIPTION
**Pull request by kuriatsu**
## Description

Initial PR for adding TSUBASA keyboard and keymaps (default and rotary encoder featured keymap)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

None 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/handwired/tsubasa/info.json [hunk 1]
```diff
@@ -0,0 +1,75 @@
+{
+    "keyboard_name": "tsubasa",
+    "url": "https://github.com/kuriatsu/TSUBASA",
+    "maintainer": "kuriatsu",
```
Comment (drashna) [inline keyboards/handwired/tsubasa/info.json:4]: You need the manufacturer name here.
```suggestion
    "maintainer": "kuriatsu",
    "manufacturer": "kuriatsu",
```

File: keyboards/handwired/tsubasa/keymaps/default/readme.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# The default keymap for tsubasa
+![keymap](https://imgur.com/wIRs6Eb)
```
Comment (drashna) [inline keyboards/handwired/tsubasa/keymaps/default/readme.md:2]: I think you want: 
```suggestion
![keymap](https://i.imgur.com/wIRs6Ebh.png)
```

File: keyboards/handwired/tsubasa/tsubasa.c [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "tsubasa.h"
```
Comment (drashna) [inline keyboards/handwired/tsubasa/tsubasa.c:17]: This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion
#include "tsubasa.h"

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by kuriatsu** [comment]
Thanks for your review @drashna

**Review by drashna — APPROVED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by kuriatsu** [comment]
Thanks @fauxpark

File: keyboards/handwired/tsubasa/config.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* key matrix size */
+#define MATRIX_ROWS 10
+#define MATRIX_COLS 6
+
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/config.h:25]:
```suggestion
```

File: keyboards/handwired/tsubasa/config.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#    define RGBLIGHT_VAL_STEP 1
+#    define RGBLIGHT_LIMIT_VAL 255 /* The maximum brightness level */
+#    define RGBLIGHT_EFFECT_BREATHING
+#    define RGBLIGHT_EFFECT_RAINBOW_MOOD
+#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+#endif
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCE 5
+
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/config.h:63]:
```suggestion
```

File: keyboards/handwired/tsubasa/rules.mk [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina
+
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/rules.mk:6]:
```suggestion
```

File: keyboards/handwired/tsubasa/keymaps/default/readme.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+## Build and Flash
+```bash
+qmk compile -kb handwired/tsubasa -km default
+```
+
+Flashing example for this keyboard:
+
+```bash
+qmk flash -kb handwired/tsubasa -km default
+```
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/keymaps/default/readme.md:13]:
```suggestion
```

File: keyboards/handwired/tsubasa/info.json [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    "manufacturer": "kuriatsu",
+    "usb": {
+        "vid": "0xFEED",
+        "pid": "0x0000",
+        "device_version": "1.0.0"
+    },
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/info.json:10]:
```suggestion
    },
    "processor": "atmega32u4",
    "bootloader": "caterina",
```

File: keyboards/handwired/tsubasa/rules.mk [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/rules.mk:15]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by kuriatsu** [comment]
@fauxpark Thanks for your review.

File: keyboards/handwired/tsubasa/config.h [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+#define MATRIX_ROW_PINS { C6, D7, E6, B4, B5 }
+#define MATRIX_COL_PINS { F6, F7, B1, B3, B2, B6 }
+
+/* COL2ROW, ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+#define SOFT_SERIAL_PIN D3  // or D1, D2, D3, E6
```
Comment (waffle87) [inline keyboards/handwired/tsubasa/config.h:39]:
```suggestion
#define MASTER_RIGHT

#define MATRIX_ROW_PINS { C6, D7, E6, B4, B5 }
#define MATRIX_COL_PINS { F6, F7, B1, B3, B2, B6 }
```

File: keyboards/handwired/tsubasa/info.json [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    "processor": "atmega32u4",
+    "bootloader": "caterina",
```
Comment (waffle87) [inline keyboards/handwired/tsubasa/info.json:12]:
```suggestion
    "bootloader": "caterina",
    "diode_direction": "COL2ROW",
    "split": {
        "enabled": true,
        "soft_serial_pin": "D3"
    },
```

File: keyboards/handwired/tsubasa/rules.mk [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = yes           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+LTO_ENABLE = yes
+
+SPLIT_KEYBOARD = yes
```
Comment (waffle87) [inline keyboards/handwired/tsubasa/rules.mk:15]:
```suggestion
```

File: keyboards/handwired/tsubasa/rules.mk [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+ENCODER_ENABLE = yes
+
+OLED_ENABLE = yes
+OLED_DRIVER = SSD1306
+WPM_ENABLE = yes
```
Comment (waffle87) [inline keyboards/handwired/tsubasa/rules.mk:20]: If OLEDs are enabled at the keyboard level, there should exist some functionality at the keyboard level as well.
eg. in a `tsubasa.c` file using `oled_task_kb()` function.

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/handwired/tsubasa/rules.mk (hunk 13 above)
Comment (kuriatsu) [reply keyboards/handwired/tsubasa/rules.mk:19]: Thanks for your review.
Is it should be at the keyboard level? not keymap level?
https://github.com/kuriatsu/qmk_firmware/blob/28ebdf6784548b015c57bf22a5e2d34ea330e50a/keyboards/handwired/tsubasa/keymaps/default/keymap.c#L51-L60
```c
#ifdef OLED_ENABLE
static void render_scrl(void) {
    static const char PROGMEM raw_scrl[] = {
        128,192,192,224, 96, 48, 48, 48, 48, 48, 48, 48,240,240,240,240,240,112,112, 56, 56, 56, 56, 24, 28, 28, 28,124,248,248,240, 64,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,128,192,192,224,112, 56, 56, 56, 56,  8,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        7, 15, 28, 24, 24, 24, 12, 12,  0,  0,192,254,255,255,255, 15, 32, 32, 48, 48, 16, 24, 24,  8, 12,140,156,254,255, 51,129,192,192,224,224,224,224,224,192,192,192,224,224,224,  0,  0,  0,  0,192,192,224, 96,  0,  0,  0,240,254,255,255,135,192,224,224,224,224,224,128,  0,  0,128,128,192,224, 96, 96,224,224,224,224,224, 96,  0,  0,  0,128,192,192,224,224,224,224,224, 64,  0,128,128,192,224, 96,224,224,224,224,224,224, 96,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,128,192,252,255,255,127, 15,  1, 16, 24,  8, 12, 12, 12,132,134,198,238,127, 59,131,  0, 30, 63, 63,124,248,240,240,  3,  1,193,253,255,127,  7,  0,  0,128,240,254,255, 63,  7,  0,128,224,254,255,127,  7,  1,  0,  0,192,252,255,127, 63,251,252,255,127,  7,  1,  0,128,192,252,255,255, 31,  1,128,192,192, 30, 63, 63,124,248,240,240,193,224,252,255,255,  7,  1,  0,128,192,252,255,255, 31,  1,128,128,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        96,112,112, 48, 56, 28, 14, 15,  3,  3,  5,  4,  4,  6,  6,  2,  2,  2,  3,  3,  1,  1,  1,  1,  7,  7, 15, 31, 31, 12, 12, 12,  6,  7,  3,  1,  0,  0, 31, 31, 15, 14,  6,  7,  3, 15, 31, 15, 15,  7,  7,  3, 63, 31, 31, 31, 30, 14, 14,  6,  7,  3,  3,  1,  0,  0, 15, 31, 31, 15,  6,  6,  3, 11, 31, 15, 15,  7,  7,  7, 15, 31, 31, 12, 12, 12,  6,  7,  3,  1,  0, 15, 31, 31, 15,  6,  6,  3, 11, 31, 15, 15,  7,  7,  3,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    };
    oled_write_raw_P(raw_scrl, sizeof(raw_scrl));
}
```

Comment (waffle87) [reply keyboards/handwired/tsubasa/rules.mk:19]: If the feature is enabled at the keyboard level, functionality for it should exist at the keyboard level as well using the `oled_task_kb()` function — same thing you've done for the encoders. That way it works for QMK Configurator users.

Comment (kuriatsu) [reply keyboards/handwired/tsubasa/rules.mk:19]: OK, I understand.
I moved it to the keyboard level.

File: keyboards/handwired/tsubasa/tsubasa.c [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    oled_write_raw_P(raw_scrl, sizeof(raw_scrl));
+}
+
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+    return rotation;
+}
+
+
+bool oled_task_user(void) {
```
Comment (waffle87) [inline keyboards/handwired/tsubasa/tsubasa.c:50]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) { return false; }
```

Comment (kuriatsu) [reply keyboards/handwired/tsubasa/tsubasa.c]: Thank you so much

**Review by waffle87 — APPROVED** [review]

File: keyboards/handwired/tsubasa/info.json [hunk 15]
```diff
@@ -0,0 +1,83 @@
+{
+    "keyboard_name": "tsubasa",
+    "url": "https://github.com/kuriatsu/TSUBASA",
+    "maintainer": "kuriatsu",
+    "manufacturer": "kuriatsu",
+    "usb": {
+        "vid": "0xFEED",
+        "pid": "0x0000",
+        "device_version": "1.0.0"
+    },
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/info.json:10]:
```suggestion
    },
    "matrix_pins": {
        "cols": ["F6", "F7", "B1", "B3", "B2", "B6"],
        "rows": ["C6", "D7", "E6", "B4", "B5"]
    },
```

File: keyboards/handwired/tsubasa/config.h [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+#define MASTER_RIGHT
+
+#define MATRIX_ROW_PINS { C6, D7, E6, B4, B5 }
+#define MATRIX_COL_PINS { F6, F7, B1, B3, B2, B6 }
+
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/config.h:23]:
```suggestion
```

File: keyboards/handwired/tsubasa/info.json [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    "processor": "atmega32u4",
+    "bootloader": "caterina",
+    "diode_direction": "COL2ROW",
+    "split": {
+        "enabled": true,
+        "soft_serial_pin": "D3"
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/info.json:16]:
```suggestion
        "soft_serial_pin": "D3",
        "encoder": {
            "right": {
                "rotary": [
                    {"pin_a": "F4", "pin_b": "F5"}
                ]
            }
        }
```

File: keyboards/handwired/tsubasa/config.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+#endif
+
+#define ENCODERS_PAD_A { }
+#define ENCODERS_PAD_B { }
+#define ENCODER_RESOLUTIONS { }
+#define ENCODERS_PAD_A_RIGHT { F4 }
+#define ENCODERS_PAD_B_RIGHT { F5 }
+#define ENCODER_RESOLUTIONS_RIGHT { 4 }
+
```
Comment (fauxpark) [inline keyboards/handwired/tsubasa/config.h:44]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Pull request merged**